internal/bgp/bgp.go

```diff
@@ -155,8 +161,21 @@ func (s *Session) connect() error {
 	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
 	defer cancel()
 
-	var d net.Dialer
-	conn, err := d.DialContext(ctx, "tcp", s.addr)
+	laddr, err := net.ResolveTCPAddr("tcp", net.JoinHostPort("0.0.0.0", "0"))
+	var conn net.Conn
+	d := TCPDialer{
+		Dialer: net.Dialer{
+			LocalAddr: laddr,
+			Timeout:   10 * time.Second,
+		},
+		AuthPassword: s.password,
+	}
+
+	d.TTL = 10
+	tcphost, portstr, err := net.SplitHostPort(s.addr)
```

Comment:
Move this into the DialTCP function. Convention in Go is that dial takes a string address. The dialer is the one that cares about breaking up the definition, so it should be the one doing stuff with the address.

This is especially true because, inside DialTCP, you immediately do `net.JoinHostPort` to reconstruct an address string, and pass it to net.ResolveTCPAddr... So you're doing a bunch of work for nothing here :)